### PR TITLE
WIP: pre-caching reconcilliation

### PR DESCRIPTION
**Pull request by vitus133**
/cc @irinamihai @browsell @imiller0 
Adds precache functionality to the reconcile loop.
This is still WIP
There are several issues/workarounds still exist in the code:
    1. Pre-cache software spec is attained form the overrides configmap and not from managed policies
    2. Pre-cache image pull spec is attained from the overrides configmap and not from the operator CSV
    3. Reconciliation logic is under debug

### CODE REVIEW

File: controllers/utils/constants.go [hunk 1]
```diff
@@ -24,3 +24,23 @@ const (
...
+	OperatorConfigOverrides    = "cluster-group-upgrade-overrides"
+	PrecacheJobNamespace       = "pre-cache"
+	PrecacheJobName            = "pre-cache"
+	PrecacheServiceAccountName = "pre-cache-agent"
+	PrecacheSpecCmName         = "pre-cache-spec"
+	JobNotFoundString          = "jobs.batch \"pre-cache\" not found"
+
+	PrecacheNotStarted         = "NotStarted"
+	PrecacheStarting           = "Starting"
+	PrecacheFailedToStart      = "FailiedToStart"
```
Comment (irinamihai) [inline controllers/utils/constants.go:40]: Typoe in "FailiedToStart"

File: controllers/operatorconfig.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// getOperatorConfigOverrides: reads user overrides to operator configuration
+//		from a configmap. An example for such an override would be the pre-cache
+//      workload image. It's usually taken from the operator CSV, but a user might want to
+//      override it in some cases
+
```
Comment (irinamihai) [inline controllers/operatorconfig.go:18]: Extra space not needed.

File: api/v1alpha1/clustergroupupgrade_types.go [hunk 3]
```diff
@@ -70,8 +70,14 @@ type ClusterGroupUpgradeSpec struct {
 	// placement rules and placement bindings are created, but clusters are not added to the placement rule.
 	// Once set to true, the clusters start being upgrades, one batch at a time.
 	//+kubebuilder:default=true
-	Enable   bool     `json:"enable,omitempty"`
-	Clusters []string `json:"clusters,omitempty"`
+	Enable bool `json:"enable,omitempty"`
+	// This field determines whether container image pre-caching will be done on all the clusters
+	// matching the selector.
+	// If repuired, the pre-caching process starts immediately on all clusters irrespectively of
```
Comment (irinamihai) [inline api/v1alpha1/clustergroupupgrade_types.go:76]: typo: "repuired"

File: api/v1alpha1/clustergroupupgrade_types.go [hunk 4]
```diff
@@ -70,8 +70,14 @@ type ClusterGroupUpgradeSpec struct {
...
+	// the value of the "enable" flag and the remediation strategy
```
Comment (irinamihai) [inline api/v1alpha1/clustergroupupgrade_types.go:77]: The UOCR doesn't have a remediationStrategy anymore, so I think we can remove the last part of the comment.

**Comment by vitus133** [comment]
Will open another one to the main branch

**Pull request closed without merging**